src/stavisky_lab_to_nwb/redis_interfaces/redisrecordingextractor.py

```diff
@@ -0,0 +1,276 @@
+            (channel_count, frames_per_entry). If channel_dim = 1,
+            then the data is assumed to originally be of shape
+            (frames_per_entry, channel_count)
+        """
+        # initialize base class
+        BaseRecordingSegment.__init__(self, time_vector=timestamps, t_start=t_start)
+
+        # Assign Redis client and check connection
+        self._client = client
+        self._client.ping()
+
+        # arg checks
+        assert channel_dim in [0, 1]
+        assert len(entry_ids) == self._client.xlen(stream_name)
+
+        # save some variables
+        self._stream_name = stream_name
+        self._data_key = bytes(data_key, "utf-8") if isinstance(data_key, str) else data_key
+        self._channel_count = channel_count
+        self._channel_dim = channel_dim
+        self._dtype = dtype
+        self._entry_ids = entry_ids
+        self._frames_per_entry = frames_per_entry
+        self._num_samples = frames_per_entry * len(entry_ids)
+
+    def get_num_samples(self) -> int:
+        """Returns number of samples in the segment"""
+        return self._num_samples
+
+    def get_traces(
+        self,
+        start_frame: Optional[int] = None,
+        end_frame: Optional[int] = None,
+        channel_indices: Optional[list] = None,
+    ) -> np.ndarray:
+        """Gets specified recording traces from Redis
+
+        Parameters
+        ----------
+        start_frame : int, optional
+            The index of the frame to start reading data from
+        end_frame : int, optional
+            The index of the frame to stop reading data from,
+            exclusive
+        channel_indices : list, optional
+            List of channel indices to retrieve data for
+
+        Returns
+        -------
+        traces : numpy.ndarray
+            A (frames, channels) array containing raw voltage
+            trace data from Redis
+        """
+        # handle None args
+        if start_frame is None:
+            start_frame = 0
+        if end_frame is None:
+            end_frame = self._num_samples
+
+        # arg check (not allowing negative indices currently)
+        assert start_frame >= 0 and start_frame < self._num_samples
+        assert end_frame > 0 and end_frame <= self._num_samples
+
+        # convert to entry number and within-entry idx
+        start_entry_idx = start_frame // self._frames_per_entry
+        start_frame_idx = start_frame % self._frames_per_entry
+        end_entry_idx = (end_frame - 1) // self._frames_per_entry  # inclusive
+
+        # read needed entries
+        stream_entries = self._client.xrange(
+            self._stream_name,
+            min=self._entry_ids[start_entry_idx],
+            max=self._entry_ids[end_entry_idx],
+        )
+
+        # loop through, convert to numpy and stack
+        traces = []
+        for entry in stream_entries:
+            entry_data = np.frombuffer(entry[1][self._data_key], dtype=self._dtype)
```

Comment:
I have a suggestion for better performance related to this block but it's contingent on one thing - is the `entry_data` here (that is, the return of `np.frombuffer`) a memory map (`np.memmap`)? Or is it an array loaded into memory?

Reply:
It's not a memory map, unfortunately. From what I understand, when reading from Redis (with `r.xrange`) the data are loaded into memory.

Reply:
OK then, that's fine - a better approach for performance (memory and computation) would be to define `traces` as a empty numpy array of the full shape, and assign it's data through slicing; e.g., do what you do to define `entry_data` but then at line 272 it would be `traces[next_slice, :] = entry_data[:, channel_indices]` - then you don't need any `np.concatenate`, which itself allocates an additional array in memory during this function and so leads to memory spikes by doubling instantaneous usage (has been known to kill processes if too large)

Only thing this depends on is knowing the full shape of the array beforehand, is that possible? Or is that way this is done dynamically with list appending?